File: src/core/zoneMapper.js
import { bboxCenter, pointInPolygon, resolveNormalizedPoints } from '../utils/zoneGeometry'

/**
 * Map observation objects to camera zones/subzones using point-in-polygon.
 *
 * @param {object} observation
 * @param {Array<object>} zones
 * @returns {Array<{ objectId: string, zones: string[], subzones: string[] }>}
 */
export function mapObservationToZones(observation, zones) {
  const objects = observation?.objects || []
  const zoneList = zones || []

  return objects.map((objectItem) => {
    const trackId = objectItem.trackId || objectItem.track_id
    const center = bboxCenter(objectItem.bbox)
    const zonesMatched = []
    const subzonesMatched = []

    zoneList.forEach((zone) => {
      const polygon = resolveNormalizedPoints(zone, {
        width: observation.frame_width || observation.frameWidth,
        height: observation.frame_height || observation.frameHeight,
      })

      if (!pointInPolygon(center, polygon)) return

      if (zone.parent_zone_id) {
        subzonesMatched.push(zone.id)
      } else {
        zonesMatched.push(zone.id)
      }
    })

    return {
      objectId: trackId,
      zones: zonesMatched,
      subzones: subzonesMatched,
    }
  })
}

/**
 * Flat list of all zone ids (root + subzone) containing an object.
 */
export function getMatchedZoneIds(mapping) {
  return [...(mapping?.zones || []), ...(mapping?.subzones || [])]
}

export function findMappingForObject(mappings, trackId) {
  return mappings.find((item) => item.objectId === trackId) || null
}

export function objectInZone(mapping, zoneId) {
  if (!mapping || !zoneId) return false
  return mapping.zones.includes(zoneId) || mapping.subzones.includes(zoneId)
}

File: backend/app/core/runtime/zone_mapper.py
from __future__ import annotations

from typing import Any

from app.utils.zone_geometry import bbox_center, point_in_polygon, resolve_normalized_points


def map_observation_to_zones(observation: dict[str, Any], zones: list[dict[str, Any]]) -> list[dict[str, Any]]:
    mappings: list[dict[str, Any]] = []
    frame_width = observation.get("frame_width") or observation.get("frameWidth")
    frame_height = observation.get("frame_height") or observation.get("frameHeight")

    for obj in observation.get("objects") or []:
        track_id = obj.get("trackId") or obj.get("track_id")
        center = bbox_center(obj["bbox"])
        zones_matched: list[str] = []
        subzones_matched: list[str] = []

        for zone in zones:
            polygon = resolve_normalized_points(
                zone.get("points") or [],
                points_format=zone.get("points_format"),
                reference_width=zone.get("reference_width"),
                reference_height=zone.get("reference_height"),
                fallback_width=frame_width,
                fallback_height=frame_height,
            )
            if not point_in_polygon(center, polygon):
                continue

            if zone.get("parent_zone_id"):
                subzones_matched.append(zone["id"])
            else:
                zones_matched.append(zone["id"])

        mappings.append(
            {
                "objectId": track_id,
                "zones": zones_matched,
                "subzones": subzones_matched,
            }
        )

    return mappings


def object_in_zone(mapping: dict[str, Any], zone_id: str) -> bool:
    return zone_id in mapping.get("zones", []) or zone_id in mapping.get("subzones", [])
